cabal-validate/src/Cli.hs

```diff
@@ -0,0 +1,433 @@
+-- | Parse CLI arguments and resolve defaults from the environment.
+
+module Cli
+  ( Opts (..)
+  , parseOpts
+  , HackageTests (..)
+  , Compiler (..)
+  , VersionParseException (..)
+  )
+where
+
+import Control.Applicative (Alternative (many, (<|>)), (<**>))
+import Control.Exception (Exception (displayException), throw)
+import Control.Monad (forM_, when)
+import Data.Data (Typeable)
+import Data.Maybe (listToMaybe)
+import qualified Data.Text as T
+import qualified Data.Text.Lazy as T (toStrict)
+import qualified Data.Text.Lazy.Encoding as T (decodeUtf8)
+import Data.Version (Version, parseVersion)
+import GHC.Conc (getNumCapabilities)
+import Options.Applicative
+  ( FlagFields
+  , Mod
+  , Parser
+  , ParserInfo
+  , auto
+  , execParser
+  , flag
+  , flag'
+  , fullDesc
+  , help
+  , helper
+  , hidden
+  , info
+  , long
+  , maybeReader
+  , option
+  , progDesc
+  , short
+  , strOption
+  , switch
+  , value
+  )
+import qualified Options.Applicative as Opt
+import System.Directory (getCurrentDirectory)
+import System.Exit (exitSuccess)
+import System.Info (arch, os)
+import System.Process.Typed (proc, readProcessStdout_)
+import Text.ParserCombinators.ReadP (readP_to_S)
+
+import ClockUtil (AbsoluteTime, getAbsoluteTime)
+import Step (Step (..), displayStep, parseStep)
+
+-- | Command-line options, resolved with context from the environment.
+data Opts = Opts
+  { verbose :: Bool
+    -- ^ Whether to display build and test output.
+  , jobs :: Int
+    -- ^ How many jobs to use when running tests.
+    --
+    -- Defaults to the number of physical cores.
```

Comment:
I’d prefer an empty line after a haddock and the following field. Otherwise, especially when haddock has more than one line, it’s a little hard to associate it right during skimming.

Reply:
Me too, but `fourmolu` removes it automatically!